src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,26 +30,152 @@ public EntityMappingImpl( final Graph mappingDescription ) {
+				globals.add(workingTriple.getSubject());
+			} else {
+				final Set<Node> newGlobalSet = new HashSet<>();
+				l2gMap.put( workingTriple.getObject(), newGlobalSet );
+				newGlobalSet.add(workingTriple.getSubject());
+			}
 		}
 	}
 
 	@Override
 	public Set<TriplePattern> applyToTriplePattern( final TriplePattern tp ) {
-		// TODO: implement this function
-		return Collections.singleton(tp);
+		final Set<Triple> workingSet = new HashSet<Triple>();
+		if(tp.asJenaTriple().getSubject().isURI()) {
+			final Set<Node> localSubjects = g2lMap.get( tp.asJenaTriple().getSubject() );
+			if( localSubjects != null ) {
+				for (final Node localSubject : localSubjects) {
+					final Triple workingTriple = new Triple(localSubject,tp.asJenaTriple().getPredicate(), tp.asJenaTriple().getObject());
+					workingSet.add(workingTriple);
+				}
+			} else {
+				workingSet.add(tp.asJenaTriple());
+			}
+		} else {
+			workingSet.add(tp.asJenaTriple());
+		}
+		final Set<TriplePattern> resultSet = new HashSet<TriplePattern>();
+			if(tp.asJenaTriple().getObject().isURI()) {
+			final Set<Node> localObjects = g2lMap.get(tp.asJenaTriple().getObject());
+			if(g2lMap.containsKey(tp.asJenaTriple().getObject())) {
+				for (final Triple workingTriple : workingSet) {
+					for (final Node localObject : localObjects) {
+						final Triple resultTriple = new Triple(workingTriple.getSubject(), workingTriple.getPredicate(), localObject);
+						resultSet.add(new TriplePatternImpl(resultTriple));
+					}
+				}
+				return resultSet;
+			} else {
+				for (final Triple workingTriple : workingSet ) {
+					resultSet.add(new TriplePatternImpl(workingTriple));
+				}
+				return resultSet;
+			}
+		} else {
+			for (final Triple workingTriple : workingSet ) {
+				resultSet.add(new TriplePatternImpl(workingTriple));
+			}
+			return resultSet;
+		}
 	}
 
 	@Override
 	public Set<SolutionMapping> applyToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		final Set<Map<Var,Node>> cartesianProduct = new HashSet<>();
+		final Binding binding = sm.asJenaBinding();
+		final Iterator<Var> it = binding.vars();
+		
+		final Var firstVar = it.next(); // Create a number of single var-node combinations to populate the set with.
+		final Node firstNode = binding.get(firstVar);
+		if (firstNode.isURI()) {
+			final Set<Node> mappedNodes = g2lMap.get(firstNode);
+			if (mappedNodes == null) { // Local Node not different from global. 
+				final Map<Var,Node> newMap = new HashMap<>();
+				newMap.put(firstVar,firstNode);
+				cartesianProduct.add(newMap);
+			} else { // Local different from global exists, use that.
+				for (Node node : mappedNodes) {
+					final Map<Var,Node> newMap = new HashMap<>();
+					newMap.put(firstVar,node);
+					cartesianProduct.add(newMap);
+				}
+			}
+		} else {
+			final Map<Var,Node> newMap = new HashMap<>();
+			newMap.put(firstVar,firstNode);
+			cartesianProduct.add(newMap);
+		}
```

Comment:
This part is almost the same as the code inside the following while loop. An easy way to avoid such code duplication is to simply initialize `cartesianProduct` to contain a single **empty** `Map`. Then, you can go directly to the while loop without any need for treating the first variable-node pair differently.